devicesetup/src/main/java/io/particle/android/sdk/devicesetup/ui/SuccessActivity.java

```diff
@@ -124,6 +125,81 @@ protected void onCreate(Bundle savedInstanceState) {
+                ParticleDevice device = particleCloud.getDevice(getIntent().getStringExtra(EXTRA_DEVICE_ID));
+                setDeviceName(device, deviceName);
+                return null;
+            }
+
+            @Override
+            public void onSuccess(@NonNull Void result) {
+                leaveActivity(context, isSuccess);
+            }
+
+            @Override
+            public void onFailure(@NonNull ParticleCloudException e) {
+                ParticleUi.showParticleButtonProgress(SuccessActivity.this, R.id.action_done, false);
+                deviceNameView.setError(getString(R.string.device_naming_failure));
+            }
+        });
+    }
+
+    private void leaveActivity(Context context, boolean isSuccess) {
+        Intent intent = NextActivitySelector.getNextActivityIntent(
+                context,
+                particleCloud,
+                SDKGlobals.getSensitiveDataStorage(),
+                new SetupResult(isSuccess, isSuccess ? DeviceSetupState.deviceToBeSetUpId : null));
+
+        // FIXME: we shouldn't do this in the lib.  looks like another argument for Fragments.
+        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP
+                | Intent.FLAG_ACTIVITY_SINGLE_TOP
+                | Intent.FLAG_ACTIVITY_CLEAR_TASK
+                | Intent.FLAG_ACTIVITY_NEW_TASK);
+        startActivity(intent);
+
+        Intent result = new Intent(DeviceSetupCompleteContract.ACTION_DEVICE_SETUP_COMPLETE)
+                .putExtra(DeviceSetupCompleteContract.EXTRA_DEVICE_SETUP_WAS_SUCCESSFUL, isSuccess);
+        if (isSuccess) {
+            result.putExtra(DeviceSetupCompleteContract.EXTRA_CONFIGURED_DEVICE_ID,
+                    DeviceSetupState.deviceToBeSetUpId);
+        }
+        LocalBroadcastManager.getInstance(context).sendBroadcast(result);
+
+        finish();
+    }
+
+    private void setDeviceName(ParticleDevice device, String deviceName) throws ParticleCloudException {
+        //Set new device name only if it changed
+        if (!device.getName().equals(deviceName)) {
+            device.setName(deviceName);
+        }
+    }
+
+    private void showDeviceName(ParticleCloud cloud) {
+        Async.executeAsync(cloud, new Async.ApiWork<ParticleCloud, ParticleDevice>() {
+            @Override
+            public ParticleDevice callApi(@NonNull ParticleCloud cloud) throws ParticleCloudException, IOException {
+                return particleCloud.getDevice(getIntent().getStringExtra(EXTRA_DEVICE_ID));
+            }
+
+            @Override
+            public void onSuccess(@NonNull ParticleDevice particleDevice) {
+                deviceNameView.setText(particleDevice.getName());
+            }
+
+            @Override
+            public void onFailure(@NonNull ParticleCloudException e) {
+                //In case setup was successful, but we cannot retrieve device naming would be a minor issue
+                deviceNameView.setVisibility(View.GONE);
```

Comment:
Does not sound correct to me.
In case of error just show no text, also don't allow user to accept an empty name.

The lib should generate a random 2 word name as in iOS.